aqt/helper.py

```diff
@@ -188,6 +313,103 @@ def getlistint(self, section: str, option: str, fallback=[]):
         return result
 
 
+def to_version(qt_ver: str) -> Version:
+    """Converts a Qt version string with dots (5.X.Y, etc) into a semantic version.
+    If the version ends in `-preview`, the version is treated as a preview release.
+    If the patch value is missing, patch is assumed to be zero.
+    If the version cannot be converted to a Version, a CliInputError is raised.
+    """
+    match = re.match(r"^(\d+)\.(\d+)(\.(\d+)|-preview)?$", qt_ver)
+    if not match:
+        raise CliInputError("Invalid version string '{}'".format(qt_ver))
+    major, minor, end, patch = match.groups()
+    is_preview = end == "-preview"
+    return Version(
+        major=int(major),
+        minor=int(minor),
+        patch=int(patch) if patch else 0,
+        prerelease=("preview",) if is_preview else None,
+    )
+
+
+def get_semantic_version(qt_ver: str, is_preview: bool) -> Optional[Version]:
+    """Converts a Qt version string (596, 512, 5132, etc) into a semantic version.
+    This makes a lot of assumptions based on established patterns:
+    If is_preview is True, the number is interpreted as ver[0].ver[1:], with no patch.
+    If the version is 3 digits, then major, minor, and patch each get 1 digit.
+    If the version is 4 or more digits, then major gets 1 digit, minor gets 2 digits
+    and patch gets all the rest.
+    As of May 2021, the version strings at https://download.qt.io/online/qtsdkrepository
+    conform to this pattern; they are not guaranteed to do so in the future.
+    """
+    if not qt_ver or any(not ch.isdigit() for ch in qt_ver):
+        return None
+    if is_preview:
+        return Version(
+            major=int(qt_ver[:1]),
+            minor=int(qt_ver[1:]),
+            patch=0,
+            prerelease=("preview",),
+        )
+    elif len(qt_ver) >= 4:
+        return Version(
+            major=int(qt_ver[:1]), minor=int(qt_ver[1:3]), patch=int(qt_ver[3:])
+        )
+    elif len(qt_ver) == 3:
+        return Version(
+            major=int(qt_ver[:1]), minor=int(qt_ver[1:2]), patch=int(qt_ver[2:])
+        )
+    elif len(qt_ver) == 2:
+        return Version(major=int(qt_ver[:1]), minor=int(qt_ver[1:2]), patch=0)
+    raise ValueError("Invalid version string '{}'".format(qt_ver))
+
+
+def request_http_with_failover(
+    base_urls: List[str], rest_of_url: str, timeout: Tuple[float, float]
+) -> str:
+    """Make an HTTP request, using one or more base urls in case the request fails.
+    If all requests fail, then re-raise the requests.exceptions.RequestException
+    that was raised by the final HTTP request.
+    Any HTTP request that resulted in a status code >= 400 will result in a RequestException.
+    """
+    for i, base_url in enumerate(base_urls):
+        try:
+            url = base_url + "/" + rest_of_url
+            r = requests.get(url, timeout=timeout)
+            r.raise_for_status()
```

Comment:
I'm not quite sure what you're asking for here. `requests.get` calls `requests.request`, documented [here](https://docs.python-requests.org/en/latest/api/#requests.request). By default, this function allows redirects (see the `allow_redirects` parameter). By the time the `requests.get` call returns, you should not ever see a redirect status code; you'll just see the response after redirection.

Regarding the timeout, every call to the `helper.request_http_with_failover` function comes from the default 'fetch http' function in ListCommand, which populates timeout with the value from Settings. I can add a default value that pulls in the Settings value directly, if that's what you're asking.

Reply:
Again, I will need some guidance as to how to proceed here. 

I think it would make a lot of sense to get rid of the raw `request.get` here, and just use `helper.getUrl`, but i would need to make `helper.getUrl` more generic so that I can set the error messages properly. If a user tries to list modules for some Qt version that doesn't exist, I don't think it would make any sense to print "Download error!" when the user isn't trying to download anything. I would make those changes in a separate PR, because they're outside the scope of this feature.